### ISSUE TITLE
[gatsby-source-wordpress] Redirect old slugs

### ISSUE THREAD

**Issue by cr101** [gatsbyjs/gatsby#1660]
@sebastienfi As you might already know, when a post slug is changed WordPress records the previous slug in the database and then it automatically notifies Google with a "HTTP response status code 301 Moved Permanently" and redirects the old post slug to the new slug. 

This is highly useful if you change the slug of a post after it's been indexed by the Google bot. Also useful redirecting users who clicked on a link that had the old slug in it.

Is it possible to create a 404 page and dynamically query the WordPress database for a post matching the slug in the url?  Or is there another way to tackle this?

**Comment by sebastienfi** [comment]
Do you know if the old slugs are exposed using the WP REST API ? If then, while building it would be possible to create a page that 301 redirects to the new slug.

**Comment by cr101** [comment]
@sebastienfi I looked at all the fields that exist for a post object in the [WordPress REST API docs](https://developer.wordpress.org/rest-api/reference/posts/) and the old slugs are not exposed. 
Will probably need to create a custom endpoint to fetch a post's old slugs.

**Comment by garnerp** [comment]
I think that old slugs are stored in wp_postmeta.  The Meta is like a key/value store, and the WP REST API docs look like they return a meta object.  It could be in there if it exists in the key/value 
store within WP (the postmeta table).

https://wordpress.stackexchange.com/questions/33361/does-wordpress-keep-track-of-a-posts-url-history-and-provide-automatic-redirect

**Comment by sebastienfi** [comment]
It would also be possible to keep record of every slug / post_id that gatsby sees for each build in a JSON file that would be read and written at node source time, then copied to the build folder because required by the 404 page. That could be a sub-plugin of `gatsby-source-wordpress`.  It would then replace the 404 page by the proper permanent redirect.

Because for SEO it's better to have a 301 than a page with 2 URLs right ?

And how to backup that file ? If we had an endpoint that could give every old slug for all the posts, we could rebuild this json file every time. I see this better than having to call the live API each time.

**Comment by cr101** [comment]
I create a small plugin to expose the old slugs in the WP REST API

```
add_action( 'rest_api_init', 'expose_old_slugs_in_api' );
 
function expose_old_slugs_in_api() {
  $post_types = get_post_types( array( 'show_in_rest' => true, 'show_in_nav_menus' => true ) ) ;

  $args = array(
    'get_callback' => 'get_old_slugs_for_api', 
    'schema' => null
  );
  
  foreach ($post_types as $type) {    
    register_rest_field( $type, 'old_slugs', $args );
  }  
}
 
function get_old_slugs_for_api( $object ) {
  //get the id of the post object array
  $post_id = $object['id'];

  //return the post meta
  return get_post_meta( $post_id, '_wp_old_slug' );
}
```

**Comment by sebastienfi** [comment]
Awesome @cr101 !
Would you like to work on that ?
Mainly in `gatsby-node.js` a page should be created for every old slug, and a special `301.js` template should be given as argument.

**Comment by KyleAMathews** [comment]
@sebastienfi we wouldn't want to do redirects in the client — that's inefficient (the user has to load the page before seeing the redirect) and would only work for browsers so bots, etc. wouldn't work.

Instead there'd be a `createRedirect` action (similar to `createPage` etc) that would store the redirect info inside Gatsby. Then plugins could use that data to create host-specific redirect logic e.g. for Apache/Nginx/Netlify/etc. to do true HTTP 301 redirects.

See this discussion https://github.com/gatsbyjs/gatsby/pull/1068#issuecomment-313893762

**Comment by sebastienfi** [comment]
Awesome!
 So, should we close this issue and focus on #1068 ?

**Comment by KyleAMathews** [comment]
It still should be kept open because once the core work is done, then this plugin would still need a PR to add support for redirects.

**Comment by sebastienfi** [comment]
What would be the best strategy to communicate this data to the host ?

For Apache, NGinx or IIS, we could generate a file (`.htaccess`, `redirect-map.conf` or `web.config`) that could contain the URL rewrite rules. Then it would be up to a custom build script to move this file where it belongs (this could not necessarily be in `/public/`).

For Apache `.htaccess`
```
Redirect /old.html      /new.html
Redirect /customers.html      /our-customers.html
Redirect /products.html       /our-products.html
```

For NGinx
```javascript
include redirect-map.conf;
server {    
    […]   
    if ($redirect_uri) {        
        return 301 $redirect_uri;   
    }
}
```
NGinx `redirect-map.conf`
```javascript
map $request_uri $redirect_uri {
    /old.html         /new.html;
    /customers.html      /our-customers.html;
    /products.html       /our-products.html;
}
```
For IIS `web.config`
```xml
<system.webServer xdt:Transform="Insert">
   <rewrite>
      <rewriteMaps>
         <rewriteMap name="StaticRewrites" defaultValue="">
            <add key="/old.html" value="/new.html" />
            <add key="/customers.html" value="/our-customers.html" />
            <add key="/products.html" value="/our-products.html" />
         </rewriteMap>
      </rewriteMaps>
      <rule name="Rewrite Rule">
         <match url=".*" />
         <conditions>
            <add input="{StaticRewrites:{REQUEST_URI}}" pattern="(.+)" />
         </conditions>
         <action type="Redirect" url="{C:1}" redirectType="Permanent" />
      </rule>
   </rewrite>
</system.webServer>
```

**Comment by KyleAMathews** [comment]
* `gatsby-plugin-apache`
* `gatsby-plugin-nginx`
* `gatsby-plugin-iis`

Each would implement `onPostBuild` and translate the redirect info into their redirect logic.

**Comment by cr101** [comment]
@sebastienfi  On larger sites you could have thousands of redirects. Apache would have to do a lot of work sifting through those redirects.
We change the post slugs in WordPress quite often.

**Comment by KyleAMathews** [comment]
@bvaughn added 1st class support for redirects at https://github.com/gatsbyjs/gatsby/pull/2191!

Someone want to add support to the gatsby-source-wordpress plugin for the new redirect API?

**Comment by KyleAMathews** [comment]
Closing this issue as it's fairly old and we have far too many open issues! If you have additional problems in this area, please open a new issue!

**Comment by b0gd4n** [comment]
any update on this?

**Comment by wardpeet** [comment]
@b0gd4n we support redirects. Feel free to re-open an issue if you have problems.

**Comment by idosius** [comment]
I don't know if this helps, but I've created a Gatsby plugin that automatically tracks page path changes during builds and creates redirects via Gatsby’s createRedirect action: 
https://www.gatsbyjs.com/plugins/gatsby-plugin-automatic-redirects/

**Comment by mehov** [comment]
Hi team,

So is this officially supported out-of-the-box by now? It would be great if this just works automatically without me having to do anything - just like it does on Wordpress.

(Apparently, the problem with the built-in Wordpress redirect is when the post URL contains category and you change it and/or its slug, which won't get tracked.)
